File: backend/app/domain/cutoff.py
from datetime import datetime, time, timedelta, timezone
from typing import NamedTuple, Optional

# Slot start times in IST (UTC+05:30)
SLOT_TIMES = {
    "Morning": time(5, 30),
    "Evening": time(17, 30)
}

class SlotAvailability(NamedTuple):
    slot: str
    delivery_date: str
    available: bool
    cutoff_time: datetime
    slot_start: datetime
    reason: Optional[str] = None

def get_slot_availability(
    delivery_date_str: str,
    slot_name: str,
    current_time: Optional[datetime] = None,
    cutoff_hours: int = 7
) -> SlotAvailability:
    """
    Pure server-authoritative cutoff calculation engine.
    Rule: currentTime <= cutoffTime -> AVAILABLE
    Exactly 7 hours before slot start is OPEN; 6 hours 59 minutes before is CLOSED.
    """
    if current_time is None:
        current_time = datetime.now(timezone.utc)
    
    # Ensure current_time is timezone-aware
    if current_time.tzinfo is None:
        current_time = current_time.replace(tzinfo=timezone.utc)

    # Parse delivery date YYYY-MM-DD
    delivery_date = datetime.strptime(delivery_date_str, "%Y-%m-%d").date()
    
    if slot_name not in SLOT_TIMES:
        raise ValueError(f"Invalid slot name: {slot_name}. Must be 'Morning' or 'Evening'.")

    slot_time = SLOT_TIMES[slot_name]
    
    # Construct exact slot start datetime in IST (+05:30)
    ist_tz = timezone(timedelta(hours=5, minutes=30))
    slot_start = datetime.combine(delivery_date, slot_time, tzinfo=ist_tz)

    # Calculate exact cutoff boundary (slot_start - 7 hours)
    cutoff_time = slot_start - timedelta(hours=cutoff_hours)

    # Convert current_time to IST for exact comparison
    current_time_ist = current_time.astimezone(ist_tz)

    # Check date boundary: Cannot book past dates
    today_ist = current_time_ist.date()
    if delivery_date < today_ist:
        return SlotAvailability(
            slot=slot_name,
            delivery_date=delivery_date_str,
            available=False,
            cutoff_time=cutoff_time,
            slot_start=slot_start,
            reason="Cannot book for past delivery dates."
        )

    # Check 7-hour cutoff boundary: currentTime <= cutoffTime -> AVAILABLE
    available = current_time_ist <= cutoff_time

    reason = None
    if not available:
        reason = (
            f"Booking for {slot_name} slot on {delivery_date_str} closed at "
            f"{cutoff_time.strftime('%I:%M %p')}. Strict {cutoff_hours}-hour prior cutoff applies."
        )

    return SlotAvailability(
        slot=slot_name,
        delivery_date=delivery_date_str,
        available=available,
        cutoff_time=cutoff_time,
        slot_start=slot_start,
        reason=reason
    )

File: src/domain/cutoff.ts
import { DeliverySlot } from '../types';

export type SlotAvailability = {
  available: boolean;
  slotStart: Date;
  cutoffTime: Date;
  reason?: string;
};

export interface TimeProvider {
  now: () => Date;
}

/**
 * Default time provider that checks localStorage for prototypeCurrentTime
 * (allowing Playwright E2E tests and prototype controls to override current time deterministically).
 */
export const defaultTimeProvider: TimeProvider = {
  now: () => {
    if (typeof window !== 'undefined' && window.localStorage) {
      const overrideISO = localStorage.getItem('prototypeCurrentTime');
      if (overrideISO) {
        const parsed = new Date(overrideISO);
        if (!isNaN(parsed.getTime())) return parsed;
      }
    }
    return new Date();
  },
};

/**
 * Calculates slot availability based on the strict 7-hour cutoff rule.
 * 
 * Rule: slotStartDateTime - currentDateTime >= 7 hours
 * If currentTime.getTime() <= cutoffTime.getTime(), the slot is available.
 * Exactly 7 hours before slot start is available. 6h 59m before is CLOSED.
 */
export function getSlotAvailability(
  deliveryDate: string, // YYYY-MM-DD
  slot: DeliverySlot,
  currentTime: Date,
  cutoffHours = 7
): SlotAvailability {
  if (!deliveryDate || !/^\d{4}-\d{2}-\d{2}$/.test(deliveryDate)) {
    return {
      available: false,
      slotStart: new Date(NaN),
      cutoffTime: new Date(NaN),
      reason: 'Invalid delivery date format',
    };
  }

  const startTimeStr = slot === 'Morning' ? '05:30' : '17:30';
  const slotStart = new Date(`${deliveryDate}T${startTimeStr}:00+05:30`);

  if (isNaN(slotStart.getTime())) {
    return {
      available: false,
      slotStart: new Date(NaN),
      cutoffTime: new Date(NaN),
      reason: 'Invalid slot date time',
    };
  }

  const cutoffTime = new Date(slotStart.getTime() - cutoffHours * 60 * 60 * 1000);
  
  // Check if delivery date itself is strictly in the past
  const startOfDayDate = new Date(`${deliveryDate}T00:00:00+05:30`);
  const startOfCurrentDay = new Date(currentTime);
  startOfCurrentDay.setHours(0, 0, 0, 0);

  if (startOfDayDate.getTime() < startOfCurrentDay.getTime()) {
    return {
      available: false,
      slotStart,
      cutoffTime,
      reason: 'Cannot select past delivery date',
    };
  }

  // currentTime <= cutoffTime -> available
  const available = currentTime.getTime() <= cutoffTime.getTime();

  return {
    available,
    slotStart,
    cutoffTime,
    reason: available ? undefined : `Booking closed (Must book at least ${cutoffHours} hours prior to slot start)`,
  };
}
